Testing our apis (#11)
Ada-Developers-Academy/core-unit-2: testing-our-apis/fixtures.md

> # Fixtures

Reviewer: Updated fixtures.md with content

| Fixture              | Shared code used to perform setup and cleanup for tests.                                     | Setup code, Cleanup code | "We're doing a lot of the same setup and cleanup in these tests. Let's refactor that into fixtures that can be shared among all of them!" |
| Dependency           | In this context, a fixture listed as one of the parameters in a test function.               | -                        | "My test was failing because I forgot to list one of its dependencies in the parameter list."                                             |
| Dependency Injection | A way for code to explicitly declare and receive the resources it needs to run successfully. | -                        | "My tests receive the resources they need because `pytest` automatically injects the dependencies."                                       |

## Creating Basic Fixtures

We create a fixture with the `@pytest.fixture` decorator.

> Let's look at a small test file. Let's run `touch test_fixtures.py` to create a new file, then paste in the following code.

Author: For the sake of making a decision for them, I may recommend being really concrete here:

```suggestion
Let's create a small test file to practice using fixtures and tests. We can create this small test file in any folder of notes. Inside our notes folder, run `touch test_fixtures.py` to create a new file, then paste in the following code.
```

Author: ```suggestion
Let's create at a small test file that we'll use to experiment for this lesson. We'll only use this test for this lesson, so we'll delete this file at the end of the lesson. Inside of our Hello Books API project, let's run `touch test_fixtures.py` to create a new file, then paste in the following code.
```

Reviewer: Thanks! Made a few edits to your comment. (tried to add my own comment to preserve your original input, but then GH wouldn't let me include my comment in the batch?)

| `@pytest.fixture`                                   | Applies the fixture decorator to the `empty_list` function. `pytest` will be able to use `empty_list` as a fixture in subsequent tests.                                                  |
| `def empty_list(): ...`                             | Declares the `empty_list` function, which returns an empty list.                                                                                                                         |
| `def test_len_of_empty_list( ... ): ...`            | Declares a test function that `pytest` will be able to detect. It begins with `test_`.                                                                                                   |
| `empty_test`                                        | This parameter matches the name of the `empty_list` fixture. When `pytest` runs this test it will first run the `empty_list` fixture and use the result as the value for this parameter. |

> We can run this test file with `pytest test_fixtures.py`. If we are not in our Hello Books API environment, we may need to create a virtual environment and install `pytest` so that the `pytest` command is available.

Author: Alternatively,

```suggestion
We can run this test file with `python3 -m pytest test_fixtures.py`.
```

Author: ```suggestion
Because we're in the Hello Books API project, let's activate our virtual environment. Then, let's run this test file with `pytest test_fixtures.py`.
```

Reviewer: We will specifically direct the reader to create this file in Hello Books API.

Reviewer: Thanks. Made some edits here too.

<br />

Generators and coroutines are powerful features in Python, but are beyond the scope of this curriculum. We will only use `yield` in the context of fixtures. But generators and coroutines can be an interesting topic for further research. Follow your curiosity!

### !end-callout

## Summary

Fixtures let us move our setup and cleanup code into shared functions. These functions are clearly marked with the `@pytest.fixture` decorator so we can be confident that those functions are used for test setup and cleanup. We can declare that a test depends on a particular fixture by listing its name as a parameter to the test.

The examples we looked at in this lesson aren't very "useful", but in real cases where we use fixtures, we can perform more complicated setup. We could initialize the resources needed to test our API endpoints, or set up or tear down a test database, as we'll see in the next lesson!

> 

Reviewer: ```suggestion

### !callout-info

## Let's Cleanup Our Test File!

We can delete the `test_fixtures.py` file from our project now! Or we can set it aside in another folder if we'd like to refer back to it later. But let's be sure to remove it from this project since it was only being used as an example for this lesson!

### !end-callout

```

## Check for Understanding

<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
### !challenge

* type: multiple-choice
* id: cb671089-b4e2-4f76-9c3f-ace9c2e88f4e
* title: Fixtures

##### !question

We want to write a test making use of a fixture. Assuming our test file imports `pytest`, which of the following correctly declares a fixture and a test that uses it?

##### !end-question

##### !options

* ```python
@pytest.fixture
def my_fixture():
    pass

def check_my_fixture(my_fixture):
    pass
```
* ```python
def my_fixture():
    pass

def test_my_fixture(my_fixture):
    pass
```
* ```python
@pytest.fixture
def my_fixture():
    pass

def test_my_fixture():
    my_fixture()
```
* ```python
@pytest.fixture
def my_fixture():
    pass

def test_my_fixture(my_fixture):
    pass
```
* ```python
@pytest.fixture
def my_fixture():
    pass

def test_my_fixture():
    pass
```

##### !end-options

##### !answer

* ```python
@pytest.fixture
def my_fixture():
    pass

def test_my_fixture(my_fixture):
    pass
```

##### !end-answer

##### !explanation

Here's an explanation for each option:

> - This option doesn't follow `pytest` naming conventions, so it will not be recognized as a test.
> - This option omits the `@pytest.fixture` decorator.
> - This option calls the fixture function directly. `pytest` disallows this and will raise an error.
> - This option is correct. It uses the decorator, follows `pytest` naming conventions, and declares the fixture as a dependency in its parameter list.
> - This option properly declares the fixture, but doesn't list it in the test parameter list.

Author: Suggest for this to be a numbered list for slightly greater ease of "following"/mapping to the above options

```suggestion
1. This option doesn't follow `pytest` naming conventions, so it will not be recognized as a test.
1. This option omits the `@pytest.fixture` decorator.
1. This option calls the fixture function directly. `pytest` disallows this and will raise an error.
1. This option is correct. It uses the decorator, follows `pytest` naming conventions, and declares the fixture as a dependency in its parameter list.
1. This option properly declares the fixture, but doesn't list it in the test parameter list.
```
